ext/process_tags.c

```diff
@@ -0,0 +1,356 @@
+        else if ((c >= 'a' && c <= 'z') ||
+                 (c >= '0' && c <= '9') ||
+                 c == '/' || c == '.' || c == '-') {
+            normalized[i] = c;
+        }
+        // Replace everything else with _
+        else {
+            normalized[i] = '_';
+        }
+    }
+    normalized[len] = '\0';
+    
+    return normalized;
+}
+
+/**
+ * Get the base name (last path segment) from a path
+ * 
+ * @param path The full path
+ * @return A newly allocated string with the base name (caller must free)
+ */
+static char *get_basename(const char *path) {
+    if (!path || !*path) {
+        return NULL;
+    }
+
+    const char *last_slash = strrchr(path, '/');
+#ifdef _WIN32
+    const char *last_backslash = strrchr(path, '\\');
+    if (last_backslash && (!last_slash || last_backslash > last_slash)) {
+        last_slash = last_backslash;
+    }
+#endif
+
+    const char *basename = last_slash ? last_slash + 1 : path;
+    
+    // If empty after slash, return NULL
+    if (!*basename) {
+        return NULL;
+    }
+
+    return strdup(basename);
+}
+
+/**
+ * Get the directory name (parent directory) from a path
+ * 
+ * @param path The full path
+ * @return A newly allocated string with the directory name (caller must free)
+ */
+static char *get_dirname(const char *path) {
+    if (!path || !*path) {
+        return NULL;
+    }
+
+    char *path_copy = strdup(path);
+    if (!path_copy) {
+        return NULL;
+    }
+
+    char *last_slash = strrchr(path_copy, '/');
+#ifdef _WIN32
+    char *last_backslash = strrchr(path_copy, '\\');
+    if (last_backslash && (!last_slash || last_backslash > last_slash)) {
+        last_slash = last_backslash;
+    }
+#endif
+
+    if (last_slash) {
+        *last_slash = '\0';
+        
+        // Get the basename of the directory
+        char *basedir = get_basename(path_copy);
+        free(path_copy);
+        return basedir;
+    }
+
+    free(path_copy);
+    return strdup(".");
+}
+
+/**
+ * Add a process tag entry
+ * 
+ * @param key The tag key (will be duplicated)
+ * @param value The tag value (will be normalized and duplicated)
+ */
+static void add_process_tag(const char *key, const char *value) {
+    if (!key || !value || process_tags.count >= MAX_PROCESS_TAGS) {
+        return;
+    }
+
+    char *normalized_value = normalize_value(value);
+    if (!normalized_value) {
+        return;
+    }
+
+    process_tags.entries[process_tags.count].key = strdup(key);
+    process_tags.entries[process_tags.count].value = normalized_value;
+    
+    if (process_tags.entries[process_tags.count].key) {
+        process_tags.count++;
+    } else {
+        free(normalized_value);
+    }
+}
+
+/**
+ * Comparison function for qsort to sort process tags by key
+ */
+static int compare_tags(const void *a, const void *b) {
+    const process_tag_entry_t *tag_a = (const process_tag_entry_t *)a;
+    const process_tag_entry_t *tag_b = (const process_tag_entry_t *)b;
```

Comment:
```suggestion
    const process_tag_entry_t *tag_a = a;
    const process_tag_entry_t *tag_b = b;
```
Needing an explicit cast of void pointers is so C++ :-D